### ISSUE TITLE
Idea: 'manually' unlock assignments for a group of students

### ISSUE THREAD

**Issue by cycomachead** [cs10/Alonzo#112]
It would be MUCH nicer to only have 1 midterm assignment in bCourses, rather than a time controlled assignment per-section. This cleans up bCourses and could simplify end-of-semester grading. 

If we could easily lock/unlock an assignment for a group of students (n=30) dynamically, this wouldn't be hard. We'd need a way to collect that many SIDs at once, and put them in. We could do ID cards, like for checkoffs OR we could potentially tie this to submitting a google form. The Google form could have a submit hook (via Google Apps JS tools) that POSTs an unlock message to the bot, which then can do whatever to unlock a midterm assignment. 

The trickiest part is actually determining lab times to unlock for, or DSP students....
DSP extensions are probably best handled manually, but lab times would be great to have an API to query.....

**Comment by aaaschmitt** [comment]
Doesn't bcourses support sections (at least for overrides it does)?

**Comment by cycomachead** [comment]
We're nice to student so we can't rely on the generated sections...too many people flip around.

**Comment by aaaschmitt** [comment]
@cycomachead  What's wrong with the password approach btw? Why aren't we just doing that now?

**Comment by aaaschmitt** [comment]
I mean TAs would have to be extra careful to reset the password once the test starts but that seems way easier than like scanning everyones ids in or even making them fill out a form of some sort

**Comment by cycomachead** [comment]
You can't password protect a single assignment though.

**Comment by aaaschmitt** [comment]
Oh I think I keep forgetting this. So we can only pw protect quizzes?

**Comment by cycomachead** [comment]
Yep, only quizzes..... 

Quizzes would be fine if we could also PW protect files or something.

**Comment by aaaschmitt** [comment]
so files can sort of be restricted. I'm checking to see if there's a good way to go about this.

**Comment by aaaschmitt** [comment]
Ok I know how we can do this.

1.) Create a single quiz for the online final (and in the future we can do this for online midterms too)
2.) Password protect the quiz using the alonzo tool
3.) We store a starter file for each section in bcourses initially leave it unpublished
4.) When a TA has section they switch their starter file to have permissions 'restrict to students with link'
5.) Once all students have the starter file they then unpublish the file again which blocks student access even via the link

And there you go. All the online tests in one quiz

**Comment by aaaschmitt** [comment]
Oh I forgot to add that in the description for the quiz we can put links to starter files for each section. The only ones that will be accessible to students are those that have the correct permissions

**Comment by cycomachead** [comment]
Yeah, ok, that was my guess. Perhaps we can publish and unpublish files from hipchat?

**Comment by aaaschmitt** [comment]
Yeah I was looking into that. I'm not sure the exact permissions we want
are available through the api though :/ because we want either unpublished
or only available by link as the two options, but idk I can look into it
more.

**Comment by cycomachead** [comment]
https://bcourses.berkeley.edu/doc/api/files.html#method.files.api_update

Not perfect, but we could probably lock and unlock files at a certain time, and set the unlock time to be `Date.now()` and the lock time to be + 60 mins. 

Actually if we name the files with a name that matches the TA name, then we can 'automatically' unlock a file when a TA sets a quiz password or does `/set midterm password` or something.

**Comment by aaaschmitt** [comment]
I think the problem I have with that is when we "unlock" the file it becomes visible to all students. So theoretically a student at home could access the starter file during that time period.

That's why I think it would be better to use the built in "only available to students with link" option. A TA when just enable this during their section time and only students who can log into the quiz would be able to see the link. Now if the api allows us to do both that would be ideal, but I couldn't figure out how that would be accomplished.

**Comment by aaaschmitt** [comment]
see this: https://guides.instructure.com/m/4152/l/41391-how-do-i-restrict-files-and-folders-to-students-in-canvas

**Comment by aaaschmitt** [comment]
What we want is to be able to turn both options restriction options on at once.

**Comment by cycomachead** [comment]
Yeah. I thought we did this before... I know we've used files once or twice and figured something out. I think we may be able to restrict a link to a folder and within a folder restrict files based on time. So as long as we don't give a link to a folder then things would be OK. I think that's what would work. 

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by aaaschmitt** [comment]
That seems like that could work

**Comment by cycomachead** [comment]
https://bcourses.berkeley.edu/courses/1268501/files/folder/Test%20restricted?preview=67116757

https://bcourses.berkeley.edu/courses/1268501/files/67116757/download?wrap=1
Yessss when linked from within a quiz it doesn't reveal the containing folder.

**Comment by cycomachead** [comment]
Wow it is really tricky to add a link to a file from the side bar from a phone. 😛

**Comment by aaaschmitt** [comment]
Can you get the folder link to show the files in the folder though? I
couldn't get that to work.

**Comment by cycomachead** [comment]
We don't want the folder link to show right? I mean only TAs should see it because the files in the folder could be accessed at time. So yeah we would still need like 10 links in the RQ. 

That's still less work than 10 separate assignments. Or maybe TAs could just BJC.link the direct URL. 

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by aaaschmitt** [comment]
Ok yeah that would work. In the assignment description we just have time controlled links to the starter file for each TA. And then students can't access these links otherwise since the folder is blocked. I think that's probably secure enough. If we wanted we could also distribute each pdfs of the prompt this way instead of handing out like a million sheets and killing lots of trees

**Comment by cycomachead** [comment]
> And then students can't access these links otherwise since the folder is blocked. I think that's probably secure enough.

The only downside is that they could easily-ish C&P the URLs, but I don't know if that's a big deal...

Also, can you see you quiz after you submit? That'd make it easier to give the other URLs to people. :/

**Comment by aaaschmitt** [comment]
Yeah but they have to be in lab to get the quiz password and submit. So
even if a friend in an earlier lab got a starter file link for a later lab
time, they would only be able to access it during that lab time and to
submit they would have to be in lab.

**Comment by cycomachead** [comment]
But once you've submitted a quiz, then I think you can see your submission which would include all links?

My worry is that it wouldn't be that hard for the list of time-controlled links to be shared since they're all in one place. Maybe it's not such a big deal, but it actually seems like it might be easier to generate temporary URLs to bCourses. (We could either use the bot or shout key)

**Comment by aaaschmitt** [comment]
The problem that you just described is why we are using time controlled links. Even if a student did manage to get a link or even all of the links ahead of time they would only be able to use it during their in lab final. You can't take the quiz at home because you won't have the password and therefore won't be able to submit. This forces a student to come to lab regardless of whether they have the links or not.

The only risk is if a student in lab were to send a friend who had collected the links the quiz password (or they could just also send them the link). Then a student could access the test from outside of the in lab final, but I believe that this risk exists in the current system anyways correct?

**Comment by cycomachead** [comment]
> The problem that you just described is why we are using time controlled links. Even if a student did manage to get a link or even all of the links ahead of time they would only be able to use it during their in lab final. You can't take the quiz at home because you won't have the password and therefore won't be able to submit. This forces a student to come to lab regardless of whether they have the links or not.

It's not about taking the final at home, it would just make it really easy outside of lab to download files.  Even though the links are time controlled, if a student who has taken a quiz can see all the links, then those links can be distributed without a quiz password.

We're never going to have a risk-free environment, but I think in lab we can limit what students do, mostly because there's a few lab assistants watching...

Apparently we can actually set quizzes to not show the quiz after submission...
https://guides.instructure.com/m/4152/l/41475-what-options-are-available-for-quizzes#!prettyPhoto

That makes things not so bad, though it still might be easier to just write a URL on the board, especially if that file with disappear shortly.

(It occurs to me that we could just upload files dropped in Hipchat to the `static` folder then delete them automatically after some time.... But that would be more development effort.)
